File: board.py
import itertools
from dataclasses import dataclass
from typing import Union

from colorama import Back, Style

EMPTY = -1

COLOR_MAPPING = {
    -1: "",
    0: Back.BLACK,
    1: Back.WHITE,
    2: Back.RED,
}


class ForbiddenMove(Exception):
    pass


@dataclass
class Board:
    """
    Stateful board of Othello game
    """
    board: list[list[int]]
    rows: int = 9
    columns: int = 9

    def __init__(self) -> None:
        super().__init__()
        self.board = [[EMPTY for _ in range(self.columns)] for _ in range(self.rows)]

    def topbot_line(self) -> str:
        return "  " + Back.GREEN + " ".join(["+"] + (["-  "] * self.columns)[:-1] + [f"- +{Style.RESET_ALL}\n"])

    def __str__(self):
        output = ""
        # column numbering
        output += "    " + "   ".join(map(str, range(self.columns))) + "\n"
        # top line
        output += self.topbot_line()
        # iterate over all but last row
        for row_num, row in enumerate(self.board):
            output += f"{row_num} {Back.GREEN}|"
            for cell in row:
                output += self.print_disc(cell) + "|"
            output += f"{Style.RESET_ALL}\n"
            if row_num != self.rows - 1:
                output += "  " + Back.GREEN + " ".join([" "] + ["-  "] * self.columns) + Style.RESET_ALL + "\n"
        # bottom line
        output += self.topbot_line()
        return output

    @staticmethod
    def print_disc(cell):
        color = COLOR_MAPPING[cell]
        return f"{color or ''}   {Back.GREEN}"

    def setup_three_players(self):
        """
        Create starting position for 3 player game
        """
        self.board[3][3] = self.board[3][5] = self.board[4][4] = 0  # white
        self.board[3][4] = self.board[5][3] = self.board[5][5] = 1  # black
        self.board[4][3] = self.board[4][5] = self.board[5][4] = 2  # red

    def place(self, row: int, column: int, player: int):
        self.validate_placing(row, column, player)
        to_flip = self.would_flip(row, column, player)
        for x, y in to_flip:
            self.board[x][y] = player
        self.board[row][column] = player

    def validate_placing(self, row: int, column: int, player: int) -> (bool, str):
        """
        Make sure move is valid.
        If not raise ForbiddenMove exception with human readable reason why the move is not allowed.
        """
        # TODO(tkarwowski)
        # outside bounds
        if row < 0 or row >= self.rows or column < 0 or column >= self.columns:
            return False, "Placement out of bounds."

        # square taken
        if self.board[row][column] != EMPTY:
            return False, "Square taken."

        # square not adjacent to any current discs
        is_adjacent = False
        for y in [row - 1, row, row + 1]:
            for x in [column - 1, column, column + 1]:
                # outside of bounds, skip
                if y < 0 or y >= self.rows or x < 0 or x >= self.columns:
                    continue
                # this is handled in square taken, but double-check
                if row == y and column == x:
                    continue
                # check if square is taken
                if self.board[y][x] != EMPTY:
                    is_adjacent = True
                    break
            if is_adjacent:
                break
        if not is_adjacent:
            return False, "New disc must be adjacent to some existing one."

        # placing must flip at least one opposing disk
        if not self.would_flip(row, column, player):
            return False, "Move won't flip any disks."

        return True, ""

    def would_flip(self, row, column, player) -> list[(int, int)]:
        flips = []
        for dy, dx in itertools.product([-1, 0, 1], [-1, 0, 1]):
            visited_discs = []
            if dy == dx == 0:
                continue
            new_row = row + dy
            new_col = column + dx
            # stop when we reach board border
            while not (new_row < 0 or new_row >= self.rows or new_col < 0 or new_col >= self.columns):
                if self.board[new_row][new_col] == player:
                    # we found outflanking disc, append visited disc to flip list
                    flips += visited_discs
                    break
                elif self.board[new_row][new_col] == EMPTY:
                    # reached blank space, no discs could be outflanked
                    break
                else:
                    # add met disc to visited discs
                    visited_discs.append((new_row, new_col))
                new_row += dy
                new_col += dx

        return flips

    def has_valid_move(self, player) -> bool:
        for row in range(self.rows):
            for column in range(self.columns):
                valid, _ = self.validate_placing(row, column, player)
                if valid:
                    return True

        return False

    def valid_moves(self, player: int):
        moves = []
        for row in range(self.rows):
            for column in range(self.columns):
                valid, _ = self.validate_placing(row, column, player)
                if valid:
                    moves.append((row, column))

        return moves

    def get_winner(self) -> Union[int, None]:
        """
        Check if game ended and if so return winning player.
        If at least two players score the same, we call a draw.

        :return: winning player number or None
        """
        scores = [0, 0, 0]
        for row in self.board:
            for cell in row:
                scores[cell] += 1

        max_score = max(scores)

        if len(list(filter(lambda x: x == max_score, scores))) > 1:
            return None  # draw

        return scores.index(max_score)

File: src/board.rs
use std::fmt::{Display, Formatter, Result};

use itertools::{iproduct, max};
use termion::color;

const EMPTY_TILE: usize = 0;
const BG_GREEN: color::Bg<color::Green> = color::Bg(color::Green);
const BG_BLACK: color::Bg<color::Black> = color::Bg(color::Black);
const BG_WHITE: color::Bg<color::White> = color::Bg(color::White);
const BG_RED: color::Bg<color::Red> = color::Bg(color::Red);
const RST_CLR: color::Bg<color::Reset> = color::Bg(color::Reset);

pub(crate) enum Move {
    Valid,
    Invalid(&'static str),
}

/// Represents game board and it's state
#[derive(Debug, Clone)]
pub(crate) struct BoardState {
    pub board: Vec<usize>,
    rows: usize,
    columns: usize,
}

impl Display for BoardState {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        // column numbering
        write!(f, " ")?;
        for i in 0..self.columns {
            write!(f, "   {}", i)?;
        }
        writeln!(f)?;
        // top line
        write!(f, "  {}+ ", BG_GREEN)?;
        for _ in 0..self.columns - 1 {
            write!(f, "-   ")?;
        }
        writeln!(f, "- +{}", RST_CLR)?;
        // iterate over all rows
        for i in 0..self.rows {
            write!(f, "{} {}|", i, BG_GREEN)?;
            for j in 0..self.columns {
                match self.board[i * self.columns + j] {
                    1 => write!(f, "{}", BG_BLACK)?,
                    2 => write!(f, "{}", BG_WHITE)?,
                    3 => write!(f, "{}", BG_RED)?,
                    _ => {}
                };
                write!(f, "   {}|", BG_GREEN)?;
            }
            writeln!(f, "{}", RST_CLR)?;
            if i != self.rows - 1 {
                write!(f, "  {}  ", BG_GREEN)?;
                for _ in 0..self.columns - 1 {
                    write!(f, "-   ")?;
                }
                writeln!(f, "-  {}", RST_CLR)?;
            }
        }
        // bottom line
        write!(f, "  {}+ ", BG_GREEN)?;
        for _ in 0..self.columns - 1 {
            write!(f, "-   ")?;
        }
        writeln!(f, "- +{}", RST_CLR)?;
        Ok(())
    }
}

impl BoardState {
    pub(crate) fn new(rows: usize, columns: usize) -> Self {
        BoardState {
            rows,
            columns,
            board: vec![0; rows * columns],
        }
    }

    /// Create starting position for 3 player game
    pub(crate) fn setup_three_players(&mut self) {
        self.board[3 * self.columns + 3] = 1;
        self.board[3 * self.columns + 5] = 1;
        self.board[4 * self.columns + 4] = 1;
        self.board[3 * self.columns + 4] = 2;
        self.board[5 * self.columns + 3] = 2;
        self.board[5 * self.columns + 5] = 2;
        self.board[4 * self.columns + 3] = 3;
        self.board[4 * self.columns + 5] = 3;
        self.board[5 * self.columns + 4] = 3;
    }

    pub(crate) fn get(&self, row: usize, column: usize) -> usize {
        self.board[row * self.columns + column]
    }

    pub(crate) fn set(&mut self, row: usize, column: usize, player: usize) {
        self.board[row * self.columns + column] = player;
    }

    pub(crate) fn place(
        &mut self,
        row: usize,
        column: usize,
        player: usize,
    ) -> Vec<((usize, usize), usize)> {
        let would_flip = self.would_flip(row, column, player);
        let mut flipped = Vec::with_capacity(would_flip.len());
        for (r, c) in would_flip {
            flipped.push(((r, c), self.board[r * self.columns + c]));
            self.board[r * self.columns + c] = player;
        }
        flipped.push(((row, column), self.board[row * self.columns + column]));
        self.board[row * self.columns + column] = player;

        flipped
    }

    fn would_flip(&self, row: usize, column: usize, player: usize) -> Vec<(usize, usize)> {
        let row = row as i32;
        let column = column as i32;
        let mut flips = Vec::new();
        // iterate over possible directions
        for (dy, dx) in iproduct!(0..3, 0..3) {
            let dy: i32 = dy - 1;
            let dx: i32 = dx - 1;
            let mut visited_discs = Vec::new();
            if dy == 0 && dx == 0 {
                continue;
            }
            let mut current_row = row + dy;
            let mut current_column = column + dx;
            // stop when we reach board border
            while !(current_row < 0
                || current_row >= self.rows as i32
                || current_column < 0
                || current_column >= self.columns as i32)
            {
                match self.board[current_row as usize * self.columns + current_column as usize] {
                    // we found outflanking disc, append visited disc to flip list
                    x if x == player => {
                        flips.append(&mut visited_discs);
                        break;
                    }
                    // reached blank space, no discs could be outflanked
                    x if x == EMPTY_TILE => break,
                    // add met disc to visited discs
                    _ => visited_discs.push((current_row as usize, current_column as usize)),
                }
                current_row += dy;
                current_column += dx;
            }
        }
        flips
    }

    /// Makes sure move is valid
    pub(crate) fn validate_placing(&self, row: usize, column: usize, player: usize) -> Move {
        // outside bounds
        if row >= self.rows || column >= self.columns {
            return Move::Invalid("Placement out of bounds.");
        }
        // square is taken
        if self.board[row * self.columns + column] != EMPTY_TILE {
            return Move::Invalid("Square taken.");
        }
        // square not adjacent to any current discs
        let mut is_adjacent = false;
        for r in row as i32 - 1..row as i32 + 2 {
            for c in column as i32 - 1..column as i32 + 2 {
                if r < 0 || r >= self.rows as i32 || c < 0 || c >= self.columns as i32 {
                    continue;
                }
                if r == row as i32 && c == column as i32 {
                    continue;
                }
                if self.board[r as usize * self.columns + c as usize] != EMPTY_TILE {
                    is_adjacent = true;
                    break;
                }
            }
            if is_adjacent {
                break;
            }
        }
        if !is_adjacent {
            return Move::Invalid("New disc must be adjacent to some existing one.");
        }

        if self.would_flip(row, column, player).is_empty() {
            return Move::Invalid("Move must flip at least one disk.");
        }

        Move::Valid
    }

    pub(crate) fn has_valid_move(&self, player: usize) -> bool {
        for row in 0..self.rows {
            for column in 0..self.columns {
                if let Move::Valid = self.validate_placing(row, column, player) {
                    return true;
                }
            }
        }
        false
    }

    pub(crate) fn valid_moves(&self, player: usize) -> Vec<(usize, usize)> {
        let mut moves = Vec::new();
        for row in 0..self.rows {
            for column in 0..self.columns {
                if let Move::Valid = self.validate_placing(row, column, player) {
                    moves.push((row, column));
                }
            }
        }
        moves
    }

    /// Check if game ended and if so return winning player.
    /// If at least two players score the same, we call a draw.
    pub(crate) fn get_winner(&self) -> usize {
        let mut scores = vec![0, 0, 0];
        for &cell in self.board.iter() {
            scores[cell as usize - 1] += 1;
        }

        let max_score = max(scores.iter()).unwrap();

        if scores.iter().filter(|&x| x == max_score).count() > 1 {
            return 0;
        }

        for (i, score) in scores.iter().enumerate() {
            if score == max_score {
                return i + 1;
            }
        }

        unreachable!();
    }
}
